### ISSUE TITLE
Relaychain pallets (WIP)

### ISSUE THREAD

**Issue by xlc** [open-web3-stack/open-runtime-module-library#264]
On hold because blocked by https://github.com/paritytech/cumulus/issues/206

---

- orml-relaychain-primitives
  - Types
    - trait UtilityCall\<Call>
      - fn batch(calls: Vec\<Call>) -> Self
      - fn as_derivative(index: u16, call: Box\<Call>) -> Self
    - trait BalancesCall\<AccountId, RelayChainBalance>
      - fn transfer(dest: AccountId, value: RelayChainBalance) -> Self
    - trait StakingCall\<AccountId, RelayChainBalance>
      - fn bond(controller: AccountId, payee: RewardDestination) -> Self
      - fn bond_extra(max_additional: RelayChainBalance) -> Self
      - fn unbond(value: RelayChainBalance) -> Self
      - fn withdraw_unbonded(num_slashing_spans: u32) -> Self
      - fn nominate(targets: Vec\<AccountId>) -> Self
      - fn rebond(value: RelayChainBalance) -> Self
    - trait DemocracyCall\<RelayChainBalance>
      - fn vote(ref_index: ReferendumIndex, vote: AccountVote\<RelayChainBalance>) -> Self
    - trait RelayChainMessage: UtilityCall + BalancesCall + StakingCall

- modules-relaychain
  - Types
    - trait RelayChainBridge
      - send(message: RelayChainMessage)
    - type StorageKey = Vec\<u8>
    - type StorageValue = Vec\<u8>
    - type StroageProof = Vec\<u8>
  - Module
    - Trait
      - MessageSender: UpwardMessageSender\<Self::RelayChainMessage>
      - RelayChainMessage: RelayChainMessage
    - fn update_state(origin, values: Vec\<StorageKey, value: StorageValue, proof: StroageProof)
      - ensure root during testing phase without verify proof
      - no ensure when real proof verification is available
